### ISSUE TITLE
Random quotes on numbers when serializing to `json`

### ISSUE THREAD

**Issue by captain-yoshi** [biojppm/rapidyaml#313]
I generate a certain YAML file which seems alright. When converting this file to JSON, some numbers are actually quoted which should not ! You can get the `yaml` and `json` datasets [here](https://drive.google.com/drive/folders/1_vMMyJ8KQ3ivZ0wreahm7wG3E3AsmTtk?usp=sharing).

Examples of random quotes poping up :
```json
"WEIRD": ["0.99356698989868164", 1.0605627298355103],
"OK":    [0, 0, 0],
"WEIRD": ["0.0064908224157989025", "0.0064917667768895626", "0.0064947893843054771"],
"OK":    [6.6227097511291504, 6.8674740791320801, 7.0403199195861816, 7.5792555809020996, 7.9916787147521973, 8.136042594909668, 8.5505847930908203, 8.701807975769043, 8.926518440246582, 8.9484291076660156, 9.0740194320678711, 9.3788108825683594, 9.406926155090332,
"WEIRD": ["0.91054189205169678", "0.98725020885467529", 1.070807933807373],
"REALLY WEIRD": 
[
  [1.5158847570419312, 1.6361792087554932],
  [1.0741721391677856, 1.1791903972625732],
  [1.4423576593399048, 1.7063977718353271],
  [1.1791903972625732],
  [1.1493504047393799, 1.1791903972625732],
  [1.1791903972625732, 1.3334760665893555],
  [1.0655292272567749, 1.4933452606201172],
  [1.0712906122207642, 1.1791903972625732],
  [1.1791903972625732, 1.830910325050354],
  ["0.13215841352939606", 1.4161584377288818],
  [1.1791903972625732, 1.5179581642150879],
  [1.1791903972625732, 1.2864601612091064],
  [1.1791903972625732, 1.6865267753601074],
  [1.1791903972625732, 1.2192368507385254],
  [1.1130030155181885, 1.5196701288223267],
  [1.0621790885925293, 1.1791903972625732]
]
```

**Comment by biojppm** [comment]
apparently the problem is with 0-leading numbers... hold on while I investigate

**Comment by biojppm** [comment]
Got it. This is a regression from the fix to #291, done in fbec340deae1256b22b9d67f41ab85a6d9da9797. It is quoting numbers like `04`, but it should not quote numbers like `0.4`.
